### ISSUE TITLE
Protocols as an argument raises incompatible type error

### ISSUE THREAD

**Issue by aliev** [microsoft/pyright#2887]
Hi there.
First example protocols as method arguments raises an error:

```python
from typing import NamedTuple, Protocol


class UserInterface(Protocol):
    first_name: str
    last_name: str


class User(NamedTuple):
    first_name: str
    last_name: str


class StorageInterface(Protocol):
    def save_user(self, user: UserInterface):
        ...


class Storage:
    def save_user(self, user: User):
        ...


class Service:
    def __init__(self, storage: StorageInterface):
        self.storage = storage


c = Service(storage=Storage())
```

```
error: Argument of type "Storage" cannot be assigned to parameter "storage" of type "StorageInterface" in function "__init__"
    "Storage" is incompatible with protocol "StorageInterface"
      "save_user" is an incompatible type
        Type "(user: User) -> None" cannot be assigned to type "(user: UserInterface) -> None"
          Parameter 1: type "UserInterface" cannot be assigned to type "User"
            "UserInterface" is incompatible with "User" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

A bit modified code, now protocols as return types:

```python
from typing import NamedTuple, Protocol


class UserInterface(Protocol):
    first_name: str
    last_name: str


class User(NamedTuple):
    first_name: str
    last_name: str


class StorageInterface(Protocol):
    def save_user(self) -> UserInterface:
        ...


class Storage:
    def save_user(self) -> User:
        ...


class Service:
    def __init__(self, storage: StorageInterface):
        self.storage = storage


c = Service(storage=Storage())
```

No errors in this second example. in mypy both examples raise an error:

```
error: Argument "storage" to "Service" has incompatible type "Storage"; expected "StorageInterface"
note: Following member(s) of "Storage" have conflicts:
note:     Expected:
note:         def save_user(self, user: UserInterface) -> Any
note:     Got:
note:         def save_user(self, user: User) -> Any
```

```
error: Argument "storage" to "Service" has incompatible type "Storage"; expected "StorageInterface"
note: Following member(s) of "Storage" have conflicts:
note:     Expected:
note:         def save_user(self) -> UserInterface
note:     Got:
note:         def save_user(self) -> User
```

Perhaps I do not fully understand the essence of the protocols, but it seemed to me that if their interfaces match, then there should be no errors.

P.S. [The same code on TypeScript.](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgKoGdoElzXk5AbwFgAoZZGYKdMAOTgFsIAuZWqUAcwG4yKANnFoNmbDtz6kAvmTIIh6dGkxQi-StRFNW7MJxC8Ni+jvH7Jc8sgQB7EBwCuCMLagAKKjVNi9BrgA0yCaiuhKGAJTq1hRgABbA6AB0XtrMyAC8mt6hUhSxCckhOpnBwj4QeciyMlagkLCIKADKrlAQOA34KCTW6HAAbhAY0O6OqmwjUJ14TRFscCAAnmQ18orKrW49Gv1DU2MTKtDzyItL0fk29ui2AhBJArZchydVNWukCsKb0APABF6FFo2zYW3aM0aSCkGjsDn0zja7hB7TBbQ6uChECiQPy8USSRRKCyRPeqysIAgAHdkM0-gCIO5KTTwYyIhEpEA)

**Comment by erictraut** [comment]
Yes, @JelleZijlstra is correct. Your workaround in the second example will work fine. Alternatively, if you change your first example such that `Storage.save_user` accepts a `UserInterface` protocol as its second parameter rather than a nominal `User` object, the error will go away.
